src/compiler/python_generator.h

```diff
@@ -48,22 +58,72 @@ struct GeneratorConfiguration {
   grpc::string beta_package_root;
 };
 
-class PythonGrpcGenerator : public grpc::protobuf::compiler::CodeGenerator {
- public:
-  PythonGrpcGenerator(const GeneratorConfiguration& config);
-  ~PythonGrpcGenerator();
+// An abstract interface representing a method.
+struct Method {
+  virtual ~Method() {}
+
+  virtual grpc::string name() const = 0;
+
+  virtual grpc::string input_type_name() const = 0;
+  virtual grpc::string output_type_name() const = 0;
+  virtual bool get_module_message_path_input(grpc::string* out) const = 0;
+  virtual bool get_module_message_path_output(grpc::string* out) const = 0;
+
+  virtual bool ClientStreaming() const = 0;
+  virtual bool ServerStreaming() const = 0; 
+};
+
+// An abstract interface representing a service
+struct Service {
+  virtual ~Service() {}
+
+  virtual grpc::string name() const = 0;
+
+  virtual int method_count() const = 0;
+  virtual std::unique_ptr<const Method> method(int i) const = 0;
```

Comment:
So, I haven't read through the code in depth, but: parsing through what I see here I'm thinking that if this return value had to be a smart pointer it'd be not a `unique_ptr` but a `shared_ptr` (or minimally a reference to a `unique_ptr`), as it doesn't make much sense for a container to offer up unique access to its items. Imagine having a vector where the index operator did this, then the expression `(vec[0], vec[0])` would *have* to result in an invalid pointer unless you hijacked the `unique_ptr` with a custom deallocator or you're returning views/proxies of some other data or somesuch (which is confusing enough that either the name of the method or documentation should make that clear at a glance). Reading the code further, [I see you're basically doing views/proxies](https://github.com/grpc/grpc/pull/7433/files#diff-353fa1b35efea06dddbf2e4e111a75dfR91).

And because I'm basically writing this response as I go along: wow, this is what we did in the C++ generator? Huh.

Unless you want to chuck some portable `unique_ptr`-like implementation (I say 'like' because I don't think we care about thread safety in this context) into the common headers (which is fine by me - dunno about everyone else), I'd say either return raw pointers with a documented contract on the caller to free (if you stick with doing views/proxies) or go with templates and policies and return wrappers by value.

... Don't actually go with templates and policies; the C++ subteam would trace the code back to my suggestion and strangle me.

Thinking back on C++ doing things similar to this: maybe unifying the approach across languages in a C++98-compliant way would be nice? Move the service/method wrappers out of the C++ generator and into some common header, add in some portable smart pointers, call it a day. @ctiller thoughts?